tests/test_database.py

```diff
@@ -4666,22 +4676,170 @@ def test_issue_163_dbc_newlines(self):
+        NODE_SHORT = "NODE_short"
+        SIG_LONG = "SIG456789_123456789_123456789_ABC"
+        SIG_SHORT = "SIG_short"
+
+        CAN = cantools.database.can
+        node_short = CAN.node.Node(NODE_SHORT, '')
+        node_long = CAN.node.Node(NODE_LONG, '')
+        sig_short = CAN.signal.Signal(name=SIG_SHORT, start=1, length=8)
+        sig_long = CAN.signal.Signal(name=SIG_LONG, start=9, length=8)
+
+        msg_long = CAN.message.Message(
+                frame_id=1,
+                name=MSG_LONG,
+                length=8,
+                signals=[sig_long],
+                senders=[NODE_LONG])
+        msg_short = CAN.message.Message(
+                frame_id=2,
+                name=MSG_SHORT,
+                length=8,
+                signals=[sig_short],
+                senders=[NODE_SHORT])
+        db = cantools.database.Database(
+                messages=[msg_short, msg_long],
+                nodes=[node_short, node_long],
+                version=''
+                )
+
+        db.refresh()
+        content = db.as_dbc_string()
+
+        # long names in special attribute lines only;
+        # definition lines with names not longer than 32 chars:
+        self.assertTrue(re.search("BO_ 1 {}: ".format(MSG_LONG[:32]), content))
+        self.assertTrue(re.search('BA_ "SystemMessageLongSymbol" BO_ 1 "{}";'.
+                                  format(MSG_LONG), content))
+        all_nodes = re.search("^BU_: (.*)$", content, flags=re.M).group(1).split()
+        self.assertTrue(NODE_LONG[:32] in all_nodes)
+        self.assertTrue(re.search('BA_ "SystemNodeLongSymbol" BU_ {} "{}";'.
+                                  format(NODE_LONG[:32], NODE_LONG),
+                                  content))
+        self.assertTrue(re.search("SG_ {} :".format(SIG_LONG[:32]), content))
+        self.assertTrue(re.search('BA_ "SystemSignalLongSymbol" SG_ 1 {} "{}";'.
+                                  format(SIG_LONG[:32], SIG_LONG),
+                                  content))
+
+        # NO long name attributes for objects with short names
+        self.assertFalse(re.search('BA_ "SystemMessageLongSymbol" BO_ 2 ',
+                                   content))
+        self.assertFalse(re.search('BA_ "SystemNodeLongSymbol" {}'.
+                                   format(NODE_SHORT), content))
+        self.assertFalse(re.search('BA_ "SystemSignalLongSymbol" SG_ 2 ',
+                                   content))
+
+        db_readback = cantools.database.load_string(content, 'dbc')
+
+        self.assertEqual(set([MSG_LONG, MSG_SHORT]),
+                         set([msg.name for msg in db_readback.messages]))
+        self.assertEqual(set([SIG_LONG, SIG_SHORT]),
+                         set([sig.name for msg in db_readback.messages
+                              for sig in msg.signals]))
+        self.assertEqual(set([NODE_LONG, NODE_SHORT]),
+                         set([node.name for node in db_readback.nodes]))
```

Comment:
Why compare dumped DBC with a reference file instead of lots of assertions?

Reply:
It's not a real reference file. `db_readback` is assigned lik that:
```
# genarate dbc string from the db object built up from scratch:
content = db.as_dbc_string()
...
# read back that string to double check that what has been created
# is wellformed and can be imported again:
db_readback = cantools.database.load_string(content, 'dbc')
```

I'll add some comments in this test case to make it clear.

Reply:
I'll probably create a reference file after the merge. Just so you know.